core/variable_map.js

```diff
@@ -164,17 +214,85 @@ Blockly.VariableMap.prototype.deleteVariable = function(variable) {
 };
 
 /**
- * Find the variable by the given name and return it. Return null if it is not
- *     found.
+ * Delete a variables by the passed in ID and all of its uses from this
+ * workspace. May prompt the user for confirmation.
+ * @param {string} id ID of variable to delete.
+ */
+Blockly.VariableMap.prototype.deleteVariableById = function(id) {
+  var variable = this.getVariableById(id);
+  if (variable) {
+    // Check whether this variable is a function parameter before deleting.
+    var variableName = variable.name;
+    var uses = this.getVariableUsesById(id);
+    for (var i = 0, block; block = uses[i]; i++) {
+      if (block.type == 'procedures_defnoreturn' ||
+        block.type == 'procedures_defreturn') {
+        var procedureName = block.getFieldValue('NAME');
+        var deleteText = Blockly.Msg.CANNOT_DELETE_VARIABLE_PROCEDURE.
+            replace('%1', variableName).
+            replace('%2', procedureName);
+        Blockly.alert(deleteText);
+        return;
+      }
+    }
+
+    var map = this;
+    if (uses.length > 1) {
+      // Confirm before deleting multiple blocks.
+      var confirmText = Blockly.Msg.DELETE_VARIABLE_CONFIRMATION.
+          replace('%1', String(uses.length)).
+          replace('%2', variableName);
+      Blockly.confirm(confirmText,
+          function(ok) {
+            if (ok) {
+              map.deleteVariableInternal_(variable, uses);
+            }
+          });
+    } else {
+      // No confirmation necessary for a single block.
+      map.deleteVariableInternal_(variable, uses);
+    }
+  } else {
+    console.warn("Can't delete non-existent variable: " + id);
+  }
+};
+
+/**
+ * Deletes a variable and all of its uses from this workspace without asking the
+ * user for confirmation.
+ * @param {!Blockly.VariableModel} variable Variable to delete.
+ * @param {!Array.<!Blockly.Block>} uses An array of uses of the variable.
+ * @private
+ */
+Blockly.VariableMap.prototype.deleteVariableInternal_ = function(variable,
+    uses) {
+  Blockly.Events.setGroup(true);
```

Comment:
Move the calls to setGroup up to the public method. Internal methods might be chained so they shouldn't change the grouping flag.

Reply:
I'd rather make this check if the group is already set and use that to decide whether to call setGroup(false).  Moving the setGroup calls out of here entirely feels error-prone.

Reply:
Adding the check like in block creation also works. I'll file a bug that we should do a pass on our public APIs to add setGroup wrappers.